app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/btle/BtLEQueue.java

```diff
@@ -252,16 +370,107 @@ private void handleDisconnected(int status) {
+
+        ScanSettings settings;
+        if(highPowerMode) {
+            settings = new ScanSettings.Builder()
+                    .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
+                    .build();
+        } else {
+            settings = new ScanSettings.Builder()
+                    .setScanMode(ScanSettings.SCAN_MODE_BALANCED)
+                    .build();
+        }
+
+        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            LOG.info("Using Android O+ BLE scanner");
+            List<ScanFilter> filters = Arrays.asList(new ScanFilter[]{new ScanFilter.Builder().build() });
+            //List<ScanFilter> filters = Arrays.asList(new ScanFilter[]{new ScanFilter.Builder().setDeviceAddress(mGbDevice.getAddress()).build()});
+            mBluetoothScanner.stopScan(getScanCallbackIntent(false));
+            mBluetoothScanner.startScan(filters, settings, getScanCallbackIntent(true));
+            // If high power mode is requested, we scan for 5 minutes
+            // and then continue scanning with lower priority (scan mode balanced) in order
+            // to conserve power.
+            if(highPowerMode) {
+                mHandler.postDelayed(mReduceBleScanIntervalRunnable, 5 * 60 * 1000);
+            }
+        }
+        else {
+            LOG.info("Using Android L-N BLE scanner");
+            List<ScanFilter> filters = Arrays.asList(new ScanFilter[]{ new ScanFilter.Builder().setDeviceAddress(mGbDevice.getAddress()).build() });
+            mBluetoothScanner.stopScan(mScanCallback);
+            mBluetoothScanner.startScan(filters, settings, mScanCallback);
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+                mHandler.postDelayed(mRestartRunnable, 25 * 60 * 1000);
+            }
+        }
+    }
+
+    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    private ScanCallback mScanCallback = new ScanCallback() {
+        @Override
+        public void onScanResult(int callbackType, ScanResult result) {
+            String deviceName = result.getDevice().getName();
+            String deviceAddress = result.getDevice().getAddress();
+
+            LOG.info("Scanner: Found: " + deviceName + " " + deviceAddress);
+            // The filter already filtered for our specific device, so it is enough to connect to it
+            mBluetoothScanner.stopScan(mScanCallback);
+            mHandler.removeCallbacks(mRestartRunnable);
+            connect();
+            setDeviceConnectionState(State.CONNECTING);
+        }
+
+        @Override
+        public void onBatchScanResults(List<ScanResult> results) {
+            for (ScanResult sr : results) {
+                LOG.info("ScanCallback.onBatchScanResults.each:" + sr.toString());
+            }
+        }
+
+        @Override
+        public void onScanFailed(int errorCode) {
+            LOG.error("ScanCallback.onScanFailed:" + errorCode);
+        }
+    };
+
     public void dispose() {
         if (mDisposed) {
```

Comment:
I think the scanning and all those runnables need to be stopped/cleared on explicit `connect()` or `disconnect()`. Or is this taken care of already?

Reply:
Nope, you are right. I'm currently testing a fix for this (I ran into this problem already).